### PR TITLE
Refactor template engine

### PR DESCRIPTION
**Pull request by louischan-oursky**

### CODE REVIEW

File: pkg/core/template/engine.go [hunk 1]
```diff
@@ -50,18 +50,30 @@ func (e *Engine) Register(spec Spec) {
...
+	templateBody, spec, err := e.resolveTemplate(templateType, resolveOptions)
 	if err != nil {
 		return
 	}
 	if spec.IsHTML {
-		return RenderHTMLTemplate(string(templateType), templateBody, context)
+		return RenderHTMLTemplate(RenderOptions{
+			Name:          string(templateType),
+			TemplateBody:  templateBody,
+			Defines:       spec.Defines,
```
Comment (kiootic) [inline pkg/core/template/engine.go]: When would `Defines` be used?

Comment (louischan-oursky) [reply pkg/core/template/engine.go]: To break the elements of Auth UI into several parts. And reuse them in the main page.

Comment (kiootic) [reply pkg/core/template/engine.go]: Can user influence the content of `Defines` list? If yes, a billion laugh attack is possible.

Comment (louischan-oursky) [reply pkg/core/template/engine.go]: > Can user influence the content of Defines list?

No. The Defines list is defined by Spec which is a compile time constant.

**Pull request merged**